features/nanostack/FEATURE_NANOSTACK/nanostack-interface/NanostackInterface.cpp

```diff
@@ -644,148 +676,215 @@ nsapi_error_t NanostackInterface::socket_bind(void *handle, const SocketAddress
+
+    nanostack_unlock();
+
+    return ret;
 }
 
 nsapi_error_t NanostackInterface::socket_listen(void *handle, int backlog)
 {
-    return NSAPI_ERROR_UNSUPPORTED;
+    //Check if socket exists
+    NanostackSocket *socket = static_cast<NanostackSocket *>(handle);
+    if (handle == NULL) {
+        MBED_ASSERT(false);
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+
+    nsapi_error_t ret = NSAPI_ERROR_OK;
+
+    nanostack_lock();
+
+    if(::socket_listen(socket->socket_id, backlog) < 0) {
+        ret = NSAPI_ERROR_PARAMETER;
+    } else {
+        socket->set_listening();
+    }
+
+    nanostack_unlock();
+
+    return ret;
 }
 
 nsapi_error_t NanostackInterface::socket_connect(void *handle, const SocketAddress &addr)
 {
     // Validate parameters
-    NanostackSocket * socket = static_cast<NanostackSocket *>(handle);
-    if (NULL == handle) {
+    NanostackSocket *socket = static_cast<NanostackSocket *>(handle);
+    nsapi_error_t ret;
+    if (handle == NULL) {
         MBED_ASSERT(false);
         return NSAPI_ERROR_NO_SOCKET;
     }
 
     if (addr.get_ip_version() != NSAPI_IPv6) {
-        return NSAPI_ERROR_UNSUPPORTED;
+        ret = NSAPI_ERROR_UNSUPPORTED;
+        goto out;
+    }
+
+    if (socket->closed()) {
+        ret = NSAPI_ERROR_NO_CONNECTION;
+        goto out;
+    }
+
+    if (socket->is_connecting()) {
+        ret = NSAPI_ERROR_ALREADY;
+        goto out;
+    }
+
+    if (socket->is_connected()) {
+        ret = NSAPI_ERROR_IS_CONNECTED;
+        goto out;
     }
 
     nanostack_lock();
 
-    nsapi_error_t ret;
     ns_address_t ns_addr;
-    int random_port = socket->is_bound() ? 0 : 1;
+
     convert_mbed_addr_to_ns(&ns_addr, &addr);
-    if (0 == ::socket_connect(socket->socket_id, &ns_addr, random_port)) {
-        socket->set_connecting(&ns_addr);
-        ret = 0;
+    if (::socket_connect(socket->socket_id, &ns_addr, 0) == 0) {
+        if (socket->proto == SOCKET_TCP) {
+            socket->set_connecting(&ns_addr);
+            ret = NSAPI_ERROR_IN_PROGRESS;
+        } else {
+            ret = NSAPI_ERROR_OK;
+        }
     } else {
         ret = NSAPI_ERROR_DEVICE_ERROR;
     }
 
     nanostack_unlock();
 
+out:
     tr_debug("socket_connect(socket=%p) sock_id=%d, ret=%i", socket, socket->socket_id, ret);
 
     return ret;
 }
 
 nsapi_error_t NanostackInterface::socket_accept(void *server, void **handle, SocketAddress *address)
 {
-    return NSAPI_ERROR_UNSUPPORTED;
-}
+    NanostackSocket * socket = static_cast<NanostackSocket *>(server);
+    NanostackSocket *accepted_sock = NULL;
+    nsapi_error_t ret;
 
-nsapi_size_or_error_t NanostackInterface::socket_send(void *handle, const void *p, nsapi_size_t size)
-{
-    // Validate parameters
-    NanostackSocket * socket = static_cast<NanostackSocket *>(handle);
-    if (NULL == handle) {
+    if (handle == NULL) {
         MBED_ASSERT(false);
         return NSAPI_ERROR_NO_SOCKET;
     }
 
-    nanostack_lock();
+    if (!socket->is_listening()) {
```

Comment:
I think we don't need to lock here. As the routines setting up the mode are ecnlosed with lock/unlock.